From 839c7f6939368fe5784058975ee95062cc88d4c3 Mon Sep 17 00:00:00 2001
From: Quentin Rameau <quinq@fifth.space>
Date: Mon, 14 Nov 2016 11:49:17 +0100
Subject: [PATCH 09/23] LICENSE: update people

---
 LICENSE | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git LICENSE LICENSE
index 4fbb67e..90121f5 100644
--- LICENSE
+++ LICENSE
@@ -1,18 +1,22 @@
 MIT/X Consortium License

 © 2006-2014 Anselm R Garbe <anselm@garbe.us>
-© 2010-2014 Hiltjo Posthuma <hiltjo@codemadness.org>
-© 2007-2011 Peter Hartlich <sgkkr at hartlich dot com>
-© 2010-2011 Connor Lane Smith <cls@lubutu.com>
 © 2006-2009 Jukka Salmi <jukka at salmi dot ch>
-© 2007-2009 Premysl Hruby <dfenze at gmail dot com>
+© 2006-2007 Sander van Dijk <a dot h dot vandijk at gmail dot com>
+© 2007-2011 Peter Hartlich <sgkkr at hartlich dot com>
 © 2007-2009 Szabolcs Nagy <nszabolcs at gmail dot com>
 © 2007-2009 Christof Musik <christof at sendfax dot de>
-© 2009 Mate Nagy <mnagy at port70 dot net>
+© 2007-2009 Premysl Hruby <dfenze at gmail dot com>
 © 2007-2008 Enno Gottox Boland <gottox at s01 dot de>
 © 2008 Martin Hurton <martin dot hurton at gmail dot com>
 © 2008 Neale Pickett <neale dot woozle dot org>
-© 2006-2007 Sander van Dijk <a dot h dot vandijk at gmail dot com>
+© 2009 Mate Nagy <mnagy at port70 dot net>
+© 2010-2016 Hiltjo Posthuma <hiltjo@codemadness.org>
+© 2010-2012 Connor Lane Smith <cls@lubutu.com>
+© 2011 Christoph Lohmann <20h@r-36.net>
+© 2015-2016 Quentin Rameau <quinq@fifth.space>
+© 2015-2016 Eric Pruitt <eric.pruitt@gmail.com>
+© 2016 Markus Teich <markus.teich@stusta.mhn.de>

 Permission is hereby granted, free of charge, to any person obtaining a
 copy of this software and associated documentation files (the "Software"),
--
2.11.1

